File: training-data/elixir/happy_number.exs
defmodule Happy do
  def happy?(n), do: loop(n, MapSet.new())

  defp loop(1, _seen), do: true

  defp loop(n, seen) do
    if MapSet.member?(seen, n) do
      false
    else
      next = n |> Integer.digits() |> Enum.map(&(&1 * &1)) |> Enum.sum()
      loop(next, MapSet.put(seen, n))
    end
  end
end

IO.inspect(Enum.filter(1..50, &Happy.happy?/1))

File: training-data/julia/happy_number.jl
square_digit_sum(n) = sum(d^2 for d in digits(n))

function is_happy(n::Integer)
    seen = Set{Int}()
    while n != 1 && !(n in seen)
        push!(seen, n)
        n = square_digit_sum(n)
    end
    return n == 1
end

println(filter(is_happy, 1:50))
